File: Prototype/BigGp/src/Model/SmallNodes/Constants/Constant.java
package Model.SmallNodes.Constants;

import Utils.Config.BigGpConfig;
import Model.BaseNodes.ForStatement.ForIncrement;
import Model.Interfaces.PointMutable;
import Model.Interfaces.Terminal;
import Model.Node;
import Model.SmallNodes.Expressions.Standart.Expression;
import Serialization.Token;

import java.util.ArrayList;
import java.util.Random;

public class Constant extends Node implements AddRandomConstant, PointMutable, Terminal {

    private static int MAX_INT;
    private static int MAX_STRING_LEN;
    private static final Random RANDOM = new Random();
    private String type;
    private String value;

    private String randomInteger(){
        return String.valueOf(RANDOM.nextInt(MAX_INT));
    }

    private String randomString(){
        return "XD I dont want to waste time for this :))";
    }

    private String randomBool(){
        if (RANDOM.nextInt(2) == 0){
            return "false";
        }

        return "true";
    }

    @Override
    public String getValue() {
        return type + " " + value;
    }

    @Override
    public void mutate() {

        if (this.getParentNode() instanceof ForIncrement){
            String oldValue = value;
            value = randomInteger();

            while (oldValue.equals(value)) {
                value = randomInteger();
            }
        } else if (this.getParentNode() instanceof Expression expression){
            expression.mutateConstantOrVar();
        }
    }

    public void addRandomConstant(RandomConstant givenConst) {
        switch (givenConst) {
            case INT -> {
                type = RandomConstant.INT.name();
                value = randomInteger();
            }
//            case STR -> {
//                type = RandomConstant.STR.name();
//                value = randomString();
//            }
//            case BOOL -> {
//                type = RandomConstant.BOOL.name();
//                value = randomBool();
//            }
//            case NULL -> {
//                type = RandomConstant.NULL.name();
//                value = null;
//            }
        }
    }

    @Override
    public String toString() {
        return value;
    }

    public ArrayList<Node> getChildrenAsNodes(){
        ArrayList<Node> nodes = new ArrayList<>();
        nodes.add(this);

        return nodes;
    }

    public Constant(Node parentNode) {
        super(parentNode, "Constant", true);

        BigGpConfig config = BigGpConfig.readConfig();
        MAX_INT = config.getMaxInt();
        MAX_STRING_LEN = config.getMaxStringLen();

        RandomConstant randomConstant = RandomConstant.randomChild();
        addRandomConstant(randomConstant);
    }

    public Constant(Node parentNode, RandomConstant givenConst) {
        super(parentNode, "Constant", true);

        BigGpConfig config = BigGpConfig.readConfig();
        MAX_INT = config.getMaxInt();
        MAX_STRING_LEN = config.getMaxStringLen();

        addRandomConstant(givenConst);
    }

    public Constant(Node parent, ArrayList<Token> tokens) {
        super(parent, "Constant", true);

        BigGpConfig config = BigGpConfig.readConfig();
        MAX_INT = config.getMaxInt();
        MAX_STRING_LEN = config.getMaxStringLen();

        Token token = tokens.remove(0);
        String[] data = token.getVALUE().split(" ");

        type = data[0];
        value = data[1];
    }

    public Constant(Node parentNode, RandomConstant givenConst, ArrayList<Token> tokens) {
        super(parentNode, "Constant", true);

        BigGpConfig config = BigGpConfig.readConfig();
        MAX_INT = config.getMaxInt();
        MAX_STRING_LEN = config.getMaxStringLen();

        Token token = tokens.remove(0);
        String[] data = token.getVALUE().split(" ");

        type = data[0];
        value = data[1];
    }
}

File: Interactive.Genetic.Programming.Environment/src/Domain/Model/Nodes.Small/Constants/Constant.cs
using CommunityToolkit.Diagnostics;
using Configuration;
using Model.Extensions;
using Model.Interfaces;
using Model.Interfaces.Evolution;
using Model.Interfaces.Generation;
using Model.Nodes.Big.For;
using Model.Nodes.Small.Expressions.Standard;
using Utils;

namespace Model.Nodes.Small.Constants;

public class Constant : Node, IConstant, IPointMutable, ITerminal
{
    private readonly int _maxInt;
    private string _type;
    private string _value;

    public string Value => $"{_type} {_value}";

    public override List<Node> ChildrenAsNodes() => [this];

    public void AddConstant(ConstantType constantType)
    {
        _type = constantType.ToString();
        _value = RandomInteger();
    }

    public void Mutate()
    {
        if (ParentNode is ForIncrement)
        {
            var oldValue = _value;
            _value = RandomInteger();

            while (oldValue == _value)
            {
                _value = RandomInteger();
            }
        }
        else if (ParentNode is StandardExpression expression)
        {
            expression.MutateConstOrVar();
        }
    }

    private string RandomInteger() => RandomService.RandomInt(_maxInt).ToString();

    public Constant(Node parentNode, IConstantConfiguration? configuration = null) 
        : base(parentNode, "Constant", true)
    {
        Guard.IsNotNull(configuration);
        _maxInt = configuration.MaxInt;
        
        var constantType = EnumExtensions.GetRandomValue<ConstantType>();
        _type = constantType.ToString();
        _value = RandomInteger();    }

    public Constant(Node parentNode, ConstantType givenConstant, IConstantConfiguration? configuration = null)
        : base(parentNode, "Constant", true)
    {
        Guard.IsNotNull(configuration);
        
        _maxInt = configuration.MaxInt;
        _type = givenConstant.ToString();
        _value = RandomInteger();    }

    public Constant(Node parentNode, IList<Token> tokens, IConstantConfiguration? configuration = null)
        : base(parentNode, "Constant", true)
    {
        Guard.IsNotNull(configuration);
        _maxInt = configuration.MaxInt;

        var token = tokens.PopFront();
        var data = token.Value?.Split(" ") ?? throw new NullReferenceException("Token value is null");
        
        _type = data[0];
        _value = data[1];
    }

    public Constant(Node parentNode, ConstantType givenConstant, IList<Token> tokens, IConstantConfiguration? configuration = null)
        : base(parentNode, "Constant", true)
    {
        Guard.IsNotNull(configuration);
        _maxInt = configuration.MaxInt;

        var data = tokens.PopFront().Value?.Split(" ") ?? throw new NullReferenceException("Token value is null");
        
        _type = data[0];
        _value = data[1];
    }

    public override string ToString() => Value;
}